src/hotspot/share/runtime/unloadableMethodHandle.inline.hpp

```diff
@@ -0,0 +1,97 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_RUNTIME_METHOD_UNLOAD_BLOCKER_HANDLE_INLINE_HPP
+#define SHARE_RUNTIME_METHOD_UNLOAD_BLOCKER_HANDLE_INLINE_HPP
+
+#include "runtime/unloadableMethodHandle.hpp"
+
+#include "memory/universe.hpp"
+#include "oops/klass.inline.hpp"
+#include "oops/oopHandle.inline.hpp"
+#include "oops/weakHandle.inline.hpp"
+
+inline UnloadableMethodHandle::UnloadableMethodHandle(Method* method) {
+  _method = method;
+  if (method != nullptr) {
+    oop obj = get_unload_blocker(method);
+    if (obj != nullptr) {
+      _weak_handle = WeakHandle(Universe::vm_weak(), obj);
+    }
+  }
+}
+
+oop UnloadableMethodHandle::get_unload_blocker(Method* method) {
+  assert(method != nullptr, "Should be");
+  InstanceKlass* ik = method->method_holder();
+  oop klass_holder = ik->klass_holder();
+  if (klass_holder != nullptr) {
+    // Normal class, return the holder that would block unloading.
+    // This would be either classloader oop for non-hidden classes,
+    // or Java mirror oop for hidden classes.
+    return klass_holder;
+  }
+
+  // Null holder, the relevant class would not be unloaded.
+  return nullptr;
+}
+
+void UnloadableMethodHandle::release() {
+  if (_method != nullptr) {
+    _method = nullptr;
+    _weak_handle.release(Universe::vm_weak());
+    _strong_handle.release(Universe::vm_global());
+  }
+}
+
+bool UnloadableMethodHandle::is_unloaded() const {
+  // Unloaded if weak handle was set, but now had been cleared by GC.
+  return !_weak_handle.is_empty() && _weak_handle.peek() == nullptr;
+}
+
+inline void UnloadableMethodHandle::block_unloading() {
+  assert(!is_unloaded(), "Pre-condition: should not be unloaded");
+
+  if (!_weak_handle.is_empty()) {
+    assert(_weak_handle.peek() != nullptr, "Should not be cleared");
+    assert(_method->method_holder()->is_loader_alive(), "Should be alive");
+    assert(_strong_handle.is_empty(), "Should be empty");
+    oop obj = get_unload_blocker(_method);
+    if (obj != nullptr) {
+      _strong_handle = OopHandle(Universe::vm_global(), obj);
+    }
+    // Release the weak handle right away, so that is_unloaded() does not touch
+    // peek() when thread is in the wrong state.
+    _weak_handle.release(Universe::vm_weak());
+  }
+
+  assert(!is_unloaded(), "Post-condition: should not be unloaded");
+}
+
+inline Method* UnloadableMethodHandle::method() const {
+  assert(!is_unloaded(), "Should not be unloaded");
```

Comment:
Assert that `block_unloading()` was called before?

Reply:
Cannot do, since lifecycle allows accessing `method()` shortly after initialization. See the new lifecycle docs. `CompilerBroker` does it now, checking that `block_unloading()` was called here would fail.